# 电商数据评估清理步骤


**整体流程**
- 读取数据
- 对数据进行评估
- 通过评估数据得到的结论去解决一个个问题
- 每解决一个问题都回过头查看问题是否解决
- 最后把清洗后的数据保存为新的csv文件

## 了解表格基本信息 
**比如**
- 分析目标\
    此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收，本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。
- 简介\
    原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。
- 数据每列的含义如下：
1. InvoiceNo：发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母"c"开头，表示这笔交易被取消。
2. StockCode：产品代码。5位数，作为产品的唯一标识符。·Description：产品名称。
3. Quantity：产品在交易中的数量。
4. InvoiceDate：发票日期和时间。交易发生的期和时间。 UnitPrice：单价。价格单位为英镑（E）。
5. CustomerID：客户编号。5位数，作为客户的唯一标识符。 Country：国家名称。客户所居住的国家的名称。

## 读取数据

In [1]:
import pandas as pd

**dtype = str,最好读取的时候都以字符串的形式读入，不然可能会使数据失真**\
**比如一个0010008的编号可能会读取成10008**

In [2]:
original_data = pd.read_csv("data.csv", encoding="latin-1",dtype = str)
original_data.head()

InvoiceNo StockCode                          Description Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER        6   
1    536365     71053                  WHITE METAL LANTERN        6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER        8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE        6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.        6   

      InvoiceDate UnitPrice CustomerID         Country  
0  12/1/2010 8:26      2.55      17850  United Kingdom  
1  12/1/2010 8:26      3.39      17850  United Kingdom  
2  12/1/2010 8:26      2.75      17850  United Kingdom  
3  12/1/2010 8:26      3.39      17850  United Kingdom  
4  12/1/2010 8:26      3.39      17850  United Kingdom

## 评估数据
在这一部分，我将对在上一部分建立的original_data这个DataFrame所包含的数据进行评估。\
评估主要从两个方面进行：**结构和内容，即整齐度和干净度**。数据的结构性问题指不符合“**每列是一个变量，每行是一个观察值，每个单元格是一个值**” \
这三个标准**，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

### 评估数据整齐度


- 随机抽出表格中的行.sample()
- 观察是否符合--**每列是一个变量，每行是一个观察值，每个单元格是一个值**

In [3]:
original_data.sample(10)

InvoiceNo StockCode                          Description Quantity  \
36428     539451     85055         FRENCH ENAMEL UTENSIL HOLDER        1   
282982    561679     21080  SET/20 RED RETROSPOT PAPER NAPKINS        12   
297866    562955     22196         SMALL HEART MEASURING SPOONS        2   
50068     540545     22195         LARGE HEART MEASURING SPOONS        3   
526294    580660     23378    PACK OF 12 50'S CHRISTMAS TISSUES       24   
520302    580278     22318        FIVE HEART HANGING DECORATION        1   
113767    545996     21677                     HEARTS  STICKERS        1   
105306    545217     20725              LUNCH BAG RED RETROSPOT        1   
399661    571283     23168         CLASSIC CAFE SUGAR DISPENSER        2   
510404    579455     23552         BICYCLE PUNCTURE REPAIR KIT         6   

             InvoiceDate UnitPrice CustomerID         Country  
36428   12/17/2010 16:59      7.62        NaN  United Kingdom  
282982   7/28/2011 18:56      0.85      15245  United Kingdom  
297866   8/11/2011 10:14      1.63        NaN  United Kingdom  
50068     1/9/2011 15:35      1.65      13451  United Kingdom  
526294   12/5/2011 13:49      0.39      16592  United Kingdom  
520302   12/2/2011 13:11      2.95      13560  United Kingdom  
113767    3/8/2011 14:12      0.85      14367  United Kingdom  
105306   2/28/2011 16:59      4.13        NaN  United Kingdom  
399661  10/16/2011 13:51      1.25      16533  United Kingdom  
510404  11/29/2011 14:14      2.08      15239  United Kingdom

### 评估数据干净度

- info()方法可以知道数据的变量个数，变量数据类型，非空数量等dataframe的一些基本信息

In [4]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   InvoiceNo    541909 non-null  object
 1   StockCode    541909 non-null  object
 2   Description  540455 non-null  object
 3   Quantity     541909 non-null  object
 4   InvoiceDate  541909 non-null  object
 5   UnitPrice    541909 non-null  object
 6   CustomerID   406829 non-null  object
 7   Country      541909 non-null  object
dtypes: object(8)
memory usage: 33.1+ MB


- 从输出结果来看，数据共有541909条观察值，而Description、CustomerID变量存在缺失值。
- 此外，InvoiceDate的数据类型应为日期，
- CustomerID的数据类型应为字符串，应当进行数据格式转换。

#### 评估缺失数据

- 提取缺失值所在列通过isnull方法，Description中存在缺失值返回ture

In [5]:
original_data["Description"].isnull() 

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Name: Description, Length: 541909, dtype: bool

- 在了解Description存在缺失值后，根据条件提取出缺失观察值。筛选为ture的行

In [6]:
original_data[original_data["Description"].isnull() ]

InvoiceNo StockCode Description Quantity      InvoiceDate UnitPrice  \
622       536414     22139         NaN       56  12/1/2010 11:52         0   
1970      536545     21134         NaN        1  12/1/2010 14:32         0   
1971      536546     22145         NaN        1  12/1/2010 14:33         0   
1972      536547     37509         NaN        1  12/1/2010 14:33         0   
1987      536549    85226A         NaN        1  12/1/2010 14:34         0   
...          ...       ...         ...      ...              ...       ...   
535322    581199     84581         NaN       -2  12/7/2011 18:26         0   
535326    581203     23406         NaN       15  12/7/2011 18:31         0   
535332    581209     21620         NaN        6  12/7/2011 18:35         0   
536981    581234     72817         NaN       27  12/8/2011 10:33         0   
538554    581408     85175         NaN       20  12/8/2011 14:06         0   

       CustomerID         Country  
622           NaN  United Kingdom  
1970          NaN  United Kingdom  
1971          NaN  United Kingdom  
1972          NaN  United Kingdom  
1987          NaN  United Kingdom  
...           ...             ...  
535322        NaN  United Kingdom  
535326        NaN  United Kingdom  
535332        NaN  United Kingdom  
536981        NaN  United Kingdom  
538554        NaN  United Kingdom  

[1454 rows x 8 columns]

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data['UnitPrice'].astype('float') > 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

- 筛选出来结果数量为o条，说明缺失Description值的数据，同时也不具备有效的UnitPrice值。
- Description表示产品名称，UnitPrice表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description Quantity  \
622       536414     22139                              NaN       56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE        1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE        2   
1445      536544     21786               POLKADOT RAIN HAT         4   
1446      536544     21787            RAIN PONCHO RETROSPOT        2   
...          ...       ...                              ...      ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT        5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE        4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN        1   
541539    581498     85174                S/4 CACTI CANDLES        1   
541540    581498       DOT                   DOTCOM POSTAGE        1   

            InvoiceDate UnitPrice CustomerID         Country  
622     12/1/2010 11:52         0        NaN  United Kingdom  
1443    12/1/2010 14:32      2.51        NaN  United Kingdom  
1444    12/1/2010 14:32      2.51        NaN  United Kingdom  
1445    12/1/2010 14:32      0.85        NaN  United Kingdom  
1446    12/1/2010 14:32      1.66        NaN  United Kingdom  
...                 ...       ...        ...             ...  
541536  12/9/2011 10:26      4.13        NaN  United Kingdom  
541537  12/9/2011 10:26      4.13        NaN  United Kingdom  
541538  12/9/2011 10:26      4.96        NaN  United Kingdom  
541539  12/9/2011 10:26     10.79        NaN  United Kingdom  
541540  12/9/2011 10:26   1714.17        NaN  United Kingdom  

[135080 rows x 8 columns]

- CustomerID表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些CustomerID缺失的交易数据仍然有效，因此保留此变量为空的观察值。
  

#### 评估重复数据

In [9]:
original_data["UnitPrice"] = original_data["UnitPrice"].astype("float")

In [10]:
original_data["Quantity"] = original_data["Quantity"].astype("float")

#### 评估不一致数据

- 不一致数据可能存在于Country变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [11]:
original_data["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


#### 评估无效或者错误数据

In [12]:
original_data.describe()

Quantity      UnitPrice
count  541909.000000  541909.000000
mean        9.552250       4.611114
std       218.081158      96.759853
min    -80995.000000  -11062.060000
25%         1.000000       1.250000
50%         3.000000       2.080000
75%        10.000000       4.130000
max     80995.000000   38970.000000

**先筛选出Quantity数值为负数的观察值**

In [13]:
original_data[original_data["Quantity"] < 0 ]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount      -1.0   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS      -1.0   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE      -12.0   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES      -24.0   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES      -24.0   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL     -11.0   
541541   C581499         M                            Manual      -1.0   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE      -5.0   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER      -1.0   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT      -5.0   

            InvoiceDate  UnitPrice CustomerID         Country  
141      12/1/2010 9:41      27.50      14527  United Kingdom  
154      12/1/2010 9:49       4.65      15311  United Kingdom  
235     12/1/2010 10:24       1.65      17548  United Kingdom  
236     12/1/2010 10:24       0.29      17548  United Kingdom  
237     12/1/2010 10:24       0.29      17548  United Kingdom  
...                 ...        ...        ...             ...  
540449   12/9/2011 9:57       0.83      14397  United Kingdom  
541541  12/9/2011 10:28     224.69      15498  United Kingdom  
541715  12/9/2011 11:57      10.95      15311  United Kingdom  
541716  12/9/2011 11:58       1.25      17315  United Kingdom  
541717  12/9/2011 11:58       1.25      17315  United Kingdom  

[10624 rows x 8 columns]

In [14]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN     -10.0  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN     -38.0  12/2/2010 14:42        0.0   
7188      536996     22712         NaN     -20.0  12/3/2010 15:30        0.0   
7189      536997     22028         NaN     -20.0  12/3/2010 15:30        0.0   
7190      536998     85067         NaN      -6.0  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check     -26.0  12/7/2011 18:36        0.0   
535335    581212     22578        lost   -1050.0  12/7/2011 18:38        0.0   
535336    581213     22576       check     -30.0  12/7/2011 18:38        0.0   
536908    581226     23090     missing    -338.0   12/8/2011 9:56        0.0   
538919    581422     23169     smashed    -235.0  12/8/2011 15:24        0.0   

       CustomerID         Country  
2406          NaN  United Kingdom  
4347          NaN  United Kingdom  
7188          NaN  United Kingdom  
7189          NaN  United Kingdom  
7190          NaN  United Kingdom  
...           ...             ...  
535333        NaN  United Kingdom  
535335        NaN  United Kingdom  
535336        NaN  United Kingdom  
536908        NaN  United Kingdom  
538919        NaN  United Kingdom  

[1336 rows x 8 columns]

In [15]:
original_data[(original_data["Quantity"] < 0) 
            & (original_data["InvoiceNo"].str[0] != "C") 
            & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当Quantity变量为负数的时候，观察值满足以下条件之一：
- InvoiceNo以C开头，表示订单被取消
- UnitPrice为0，说明单价为0英镑\

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将Quantity为负数的观察值删除。

**接下来筛选出UnitPrice为负数的观察值**

In [16]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt       1.0  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt       1.0  8/12/2011 14:52   

        UnitPrice CustomerID         Country  
299983  -11062.06        NaN  United Kingdom  
299984  -11062.06        NaN  United Kingdom

## 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：
- InvoiceDate变量的数据类型转换为为日期时间
- 把CustomerID~变量的数据类型转换为字符串 Description变量缺失的观察值删除
- 把Country变量值“USA替换为United States”
- 把Country变量值"UK”，"U.K."替换为"United Kingdom”
- 把Quantity变量值为负数的观察值删除
- 把UnitPrice变量值为负数的观察值删除

In [17]:
cleaned_data = original_data.copy()
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER       6.0   
1         536365     71053                  WHITE METAL LANTERN       6.0   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER       8.0   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS      12.0   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL        6.0   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL        4.0   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE       4.0   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT        3.0   

            InvoiceDate  UnitPrice CustomerID         Country  
0        12/1/2010 8:26       2.55      17850  United Kingdom  
1        12/1/2010 8:26       3.39      17850  United Kingdom  
2        12/1/2010 8:26       2.75      17850  United Kingdom  
3        12/1/2010 8:26       3.39      17850  United Kingdom  
4        12/1/2010 8:26       3.39      17850  United Kingdom  
...                 ...        ...        ...             ...  
541904  12/9/2011 12:50       0.85      12680          France  
541905  12/9/2011 12:50       2.10      12680          France  
541906  12/9/2011 12:50       4.15      12680          France  
541907  12/9/2011 12:50       4.15      12680          France  
541908  12/9/2011 12:50       4.95      12680          France  

[541909 rows x 8 columns]

In [18]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

In [19]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"] 
# 在pandas中，python字符串表示为object，所以这个sh位置转换成功

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

In [20]:
cleaned_data.dropna(subset=["Description"], inplace=True)
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER       6.0   
1         536365     71053                  WHITE METAL LANTERN       6.0   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER       8.0   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS      12.0   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL        6.0   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL        4.0   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE       4.0   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT        3.0   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[540455 rows x 8 columns]

In [21]:
cleaned_data["Description"].isnull().sum()

np.int64(0)

In [22]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA":"United States"})

In [29]:
len(cleaned_data[cleaned_data["Country"]=="USA"])

0

In [33]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK":"United Kingdom","U.K.":"United Kingdom"})
print(len(cleaned_data[cleaned_data["Country"]=="UK"]))
print(len(cleaned_data[cleaned_data["Country"]=="U.K."]))

0
0


In [35]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] > 0]
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

In [37]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据
完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为data_cleaned.csv。

In [40]:
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER       6.0   
1         536365     71053                  WHITE METAL LANTERN       6.0   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER       8.0   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS      12.0   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL        6.0   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL        4.0   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE       4.0   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT        3.0   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[530691 rows x 8 columns]

In [42]:
cleaned_data.to_csv("data_cleaned.csv",index = False)